File: app/src/main/java/com/imperium/academio/ClassRegister.java
package com.imperium.academio;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.imperium.academio.databinding.ActivityClassRegisterBinding;
import com.imperium.academio.fireclass.ClassHelperClass;
import com.imperium.academio.ui.adapters.ClassRegisterRvAdapter;
import com.imperium.academio.ui.fragment.ClassDialogFragment;
import com.imperium.academio.ui.model.ClassRegisterRvModel;

import java.util.ArrayList;
import java.util.List;

public class ClassRegister extends AppCompatActivity implements ClassDialogFragment.SubmitListener {
    ActivityClassRegisterBinding binding;
    private FirebaseAuth firebaseAuth;
    DatabaseReference classes;
    String userId;
    List<ClassRegisterRvModel> cList;
    ClassRegisterRvAdapter registerRvAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_class_register);
        firebaseAuth = FirebaseAuth.getInstance();
        classes = FirebaseDatabase.getInstance().getReference("class");

        // Get user from FirebaseAuth
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (user == null || user.getDisplayName() == null) {
            startActivity(new Intent(this, Login.class));
            finish();
            return;
        }
        userId = CustomUtil.SHA1(user.getDisplayName());
        cList = (cList == null) ? new ArrayList<>() : cList;

        // Create recycler adapter for class list
        registerRvAdapter = new ClassRegisterRvAdapter(ClassRegister.this, cList);
        binding.classRegister.setHasFixedSize(true);
        binding.classRegister.setLayoutManager(new LinearLayoutManager(
                ClassRegister.this, LinearLayoutManager.VERTICAL, false
        ));

        // Attach OnClick listener to all classes in class list
        registerRvAdapter.setOnItemClickListener((itemView, position) -> {
            Toast.makeText(
                    ClassRegister.this,
                    "Opening class: " + cList.get(position).getName(),
                    Toast.LENGTH_SHORT
            ).show();
            // Open the class and finish the activity
            Intent intent = new Intent(ClassRegister.this, MainMenu.class);
            intent.putExtra("classId", cList.get(position).getKey());
            intent.putExtra("userId", userId);
            startActivity(intent);
        });


        // Attach recycler adapter for class list
        binding.classRegister.setAdapter(registerRvAdapter);

        // Create and Attach refresh listener
        binding.classRegisterSwipe.setOnRefreshListener(this::refresh);

        // Create and Attach logout button listener
        binding.btnLogout.setOnClickListener(view -> {
            // Goto Login activity after signout
            firebaseAuth.signOut();
            startActivity(new Intent(ClassRegister.this, Login.class));
            finish();
        });

        // Create and Attach register button listener
        binding.btnClassRegister.setOnClickListener(view -> {
            // Create a popup menu to choose login or create
            PopupMenu popup = new PopupMenu(ClassRegister.this, binding.btnClassRegister);
            popup.getMenuInflater().inflate(R.menu.new_class_choice, popup.getMenu());

            // Create Onclick listener for popup items
            popup.setOnMenuItemClickListener(item -> {
                // If button pressed is to join a class buttonType is true,
                // else it is false meaning to create a class.
                boolean buttonType = getString(R.string.join_class).equals(item.getTitle().toString());

                ClassDialogFragment fragment = ClassDialogFragment.newInstance(buttonType);
                fragment.show(getSupportFragmentManager(), "addClass");
                return true;
            });

            // Show the popup menu
            popup.show();
        });

        // Refresh the class list
        refresh();
    }

    // Function to perform when refresh is called
    private void refresh() {
        binding.classRegisterSwipe.setRefreshing(true);
        refreshClasses();
        // remove refreshing animation after 10 seconds
        new Handler().postDelayed(() -> binding.classRegisterSwipe.setRefreshing(false), 10000);
    }


    @Override
    public void onSubmit(boolean buttonType, String classname, String teacherName) {
        // If button pressed is to join join class
        if (buttonType) {
            ClassHelperClass classObject = new ClassHelperClass(classname, CustomUtil.SHA1(teacherName));
            joinClass(classObject);
        } else {
            // Button pressed is to create class
            ClassHelperClass classObject = new ClassHelperClass(classname, userId);
            createClass(classObject);
        }
        binding.classRegister.postDelayed(this::refreshClasses, 1000);

    }

    // Function to create a new class
    private void createClass(@NonNull ClassHelperClass classObject) {
        String key = classObject.generateKey();
        classes.child(key).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                String message;
                // Check if class exists already
                if (snapshot.exists() && snapshot.getValue(ClassHelperClass.class) != null) {
                    message = "This class already exist!";
                } else {
                    // Create class and add to classes
                    classes.child(key).setValue(classObject);

                    // Add the class to user in DB
                    FirebaseDatabase.getInstance().getReference("users/" + userId)
                            .child("classes/" + key).setValue(classObject.className);

                    message = "Creating a new class: " + classObject.className;
                }
                Toast.makeText(ClassRegister.this, message, Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });
    }

    // Function to join an existing class
    private void joinClass(@NonNull ClassHelperClass classObject) {
        String key = classObject.generateKey();
        classes.child(key).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                ClassHelperClass classFromDb = snapshot.getValue(ClassHelperClass.class);
                String message;
                // Check if class exists or not
                if (!snapshot.exists() || classFromDb == null) {
                    message = "Class Doesn't Exist!";
                } else {
                    // Add user to class in DB
                    if (!userId.equals(classFromDb.teacherId))
                        classes.child(key).child("students/" + userId).setValue(true);

                    // Add class to user in DB
                    FirebaseDatabase.getInstance().getReference("users/" + userId)
                            .child("classes/" + key)
                            .setValue(classObject.className);

                    message = "Adding class: " + classFromDb.className;
                }
                Toast.makeText(ClassRegister.this, message, Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });
    }

    // Function to refresh class list
    private void refreshClasses() {
        DatabaseReference userClasses = FirebaseDatabase.getInstance().getReference("users/" + userId + "/classes");
        userClasses.orderByValue().addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                // Check if user has joined any classes
                if (snapshot.exists()) {
                    // Clear list and add all classes
                    cList.clear();
                    for (DataSnapshot c : snapshot.getChildren()) {
                        // Add each class to the class list
                        ClassRegisterRvModel elt = new ClassRegisterRvModel(c.getKey(), c.getValue(String.class));
                        if (!cList.contains(elt))
                            cList.add(elt);
                    }
                    // Notify change to recycler view
                    binding.classRegister.post(() -> registerRvAdapter.notifyDataSetChanged());
                    binding.classRegisterSwipe.setRefreshing(false);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                binding.classRegisterSwipe.setRefreshing(false);
            }
        });
    }
}

File: app/src/main/java/com/imperium/academio/ClassRegister.kt
package com.imperium.academio

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import com.imperium.academio.databinding.ActivityClassRegisterBinding
import com.imperium.academio.fireclass.ClassHelperClass
import com.imperium.academio.ui.adapters.ClassRegisterRvAdapter
import com.imperium.academio.ui.fragment.ClassDialogFragment
import com.imperium.academio.ui.model.ClassRegisterRvModel
import java.util.*

class ClassRegister : AppCompatActivity(), ClassDialogFragment.ClassSubmitListener {
    private lateinit var binding: ActivityClassRegisterBinding

    private lateinit var classes: DatabaseReference
    private lateinit var firebaseAuth: FirebaseAuth
    private lateinit var userClasses: DatabaseReference

    private lateinit var registerRvAdapter: ClassRegisterRvAdapter
    private var cList: MutableList<ClassRegisterRvModel> = ArrayList()

    private lateinit var userId: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        firebaseAuth = FirebaseAuth.getInstance()

        binding = ActivityClassRegisterBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Get user from FirebaseAuth
        val user: FirebaseUser? = firebaseAuth.currentUser
        if (user == null || user.displayName == null) {
            startActivity(Intent(this, Login::class.java))
            return finish()
        }
        userId = CustomUtil.SHA1(user.displayName.toString())

        classes = FirebaseDatabase.getInstance().getReference("class")
        userClasses = FirebaseDatabase.getInstance().getReference("users/$userId/classes")

        // Create recycler adapter for class list
        registerRvAdapter = ClassRegisterRvAdapter(this@ClassRegister, cList)

        // Attach OnClick listener to all classes in class list
        registerRvAdapter.setOnItemClickListener(
                object : ClassRegisterRvAdapter.OnItemClickListener {
                    override fun onItemClick(itemView: View?, position: Int) {
                        toast("Opening class: " + cList[position].name)
                        // Open the class and finish the activity
                        startActivity(Intent(this@ClassRegister, MainMenu::class.java).let
                        { intent ->
                            intent.putExtra("userId", userId)
                            intent.putExtra("classId", cList[position].key)
                        })
                        finish()
                    }
                }
        )

        binding.classRegister.apply {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(
                    this@ClassRegister, LinearLayoutManager.VERTICAL, false)
            adapter = registerRvAdapter
        }


        // Create and Attach refresh listener
        binding.classRegisterSwipe.setOnRefreshListener { refresh() }

        // Create and Attach logout button listener
        binding.btnLogout.setOnClickListener {
            // Goto Login activity after signout
            firebaseAuth.signOut()
            startActivity(Intent(this@ClassRegister, Login::class.java))
            finish()
        }

        // Create and Attach register button listener
        binding.btnClassRegister.setOnClickListener {
            // Create a popup menu to choose login or create
            val popup = PopupMenu(this@ClassRegister, binding.btnClassRegister)
            popup.menuInflater.inflate(R.menu.new_class_choice, popup.menu)

            // Create Onclick listener for popup items
            popup.setOnMenuItemClickListener { item: MenuItem ->
                // If button pressed is to join a class buttonType is true,
                // else it is false meaning to create a class.
                val buttonType = getString(R.string.join_class) == item.title.toString()
                val fragment = ClassDialogFragment.newInstance(buttonType)
                fragment.show(supportFragmentManager, "addClass")
                true
            }

            // Show the popup menu
            popup.show()
        }

        // Refresh the class list
        refresh()
    }

    // Function to perform when refresh is called
    private fun refresh() {
        binding.classRegisterSwipe.apply {
            isRefreshing = true
            // remove refreshing animation after 10 seconds
            postDelayed({ isRefreshing = false }, 10000)
        }
        refreshClasses()
    }

    override fun onSubmit(buttonType: Boolean, classname: String, teacherName: String) {
        // Join or create class according to menu item chosen
        if (buttonType) joinClass(classname, CustomUtil.SHA1(teacherName))
        else createClass(classname)

        binding.classRegister.postDelayed({ refreshClasses() }, 1000)
    }

    // Function to create a new class
    private fun createClass(classname: String) {
        val classObject = ClassHelperClass(classname, userId)
        val key = classObject.generateKey()

        classes.child(key).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {

                // Check if class exists already
                if (!snapshot.exists() || snapshot.getValue(ClassHelperClass::class.java) == null) {
                    // Create class and add to classes
                    classes.child(key).setValue(classObject)

                    // Add the class to user in DB
                    FirebaseDatabase.getInstance().getReference("users/$userId")
                            .child("classes/$key").setValue(classObject.className)

                    toast("Creating a new class: " + classObject.className)
                } else {
                    toast("This class already exist!")
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    // Function to join an existing class
    private fun joinClass(classname: String, teacherId: String) {
        val tempClass = ClassHelperClass(classname, teacherId)
        val key = tempClass.generateKey()

        classes.child(key).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val instance = snapshot.getValue(ClassHelperClass::class.java)
                // Check if class exists or not
                if (snapshot.exists() && instance != null) {
                    // Add user to class in DB
                    if (userId != instance.teacherId) {
                        classes.child("$key/students/$userId").setValue(true)
                    }

                    // Add class to user in DB
                    userClasses.child(key).setValue(tempClass.className)
                    toast("Adding class: " + instance.className)
                } else {
                    toast("Class Doesn't Exist!")
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    // Function to refresh class list
    private fun refreshClasses() {
        userClasses.orderByValue().addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                // Check if user has joined any classes
                if (snapshot.exists()) {
                    // Clear list and add all classes
                    cList.clear()
                    for (c in snapshot.children) {
                        // Add each class to the class list
                        val elt = ClassRegisterRvModel(c.key, c.getValue(String::class.java))
                        if (!cList.contains(elt)) cList.add(elt)
                    }
                    // Notify change to recycler view
                    binding.classRegister.post { registerRvAdapter.notifyDataSetChanged() }
                    binding.classRegisterSwipe.isRefreshing = false
                }
            }

            override fun onCancelled(error: DatabaseError) {
                binding.classRegisterSwipe.isRefreshing = false
            }
        })
    }

    fun Context.toast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }
}